sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ContextualServer.java

```diff
@@ -0,0 +1,143 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugin.typescript;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import org.sonar.api.Startable;
+import org.sonar.api.batch.InstantiationStrategy;
+import org.sonar.api.config.Configuration;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.plugin.typescript.executable.ExecutableBundle;
+import org.sonar.plugin.typescript.executable.ExecutableBundleFactory;
+import org.sonarsource.api.sonarlint.SonarLintSide;
+
+@InstantiationStrategy("PER_PROCESS")
+@SonarLintSide
+public class ContextualServer implements Startable {
+
+  // SonarLint should pass in this property an absolute path to the directory containing TypeScript dependency
+  private static final String TYPESCRIPT_DEPENDENCY_LOCATION_PROPERTY = "sonar.typescript.internal.typescriptLocation";
+
+  private static final Logger LOG = Loggers.get(ContextualServer.class);
+
+  private Configuration configuration;
+  private ExecutableBundleFactory bundleFactory;
+  private static AtomicReference<Process> serverProcess = new AtomicReference<>();
+
+  public ContextualServer(ExecutableBundleFactory bundleFactory) {
+    this(null, bundleFactory);
+  }
+
+  public ContextualServer(Configuration configuration, ExecutableBundleFactory bundleFactory) {
+    this.configuration = configuration;
+    this.bundleFactory = bundleFactory;
+  }
+
+  @Override
+  public void start() {
+    LOG.info("Starting SonarTS Server");
+
+    if (serverProcess.get() != null && serverProcess.get().isAlive()) {
+      LOG.warn("Skipping SonarTS Server start, already running");
+      return;
+    }
+
+    if (configuration == null) {
+      LOG.warn("Skipping SonarTS Server start due to missing configuration");
+      return;
+    }
+    
+    startSonarTSServer();
+  }
+
+  private void startSonarTSServer() {
+    ExecutableBundle bundle = bundleFactory.createAndDeploy(getServerDir(), configuration);
+
+    ProcessBuilder processBuilder = new ProcessBuilder(bundle.getSonarTSServerCommand().commandLineTokens());
+    setNodePath(processBuilder);
+
+    try {
+      processBuilder.inheritIO();
+      serverProcess.set(processBuilder.start());
+      //TODO find a better way to wait until the server is up
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=sonarlint-integration-poc&issues=AWD5K7q0EFfjxkQcPcK6&open=AWD5K7q0EFfjxkQcPcK6)

[ISSUE_KEY]: AWD5K7q0EFfjxkQcPcK6